### PR TITLE
Update SecureDrop workstation to support 4.2 and use an install-time base template.

### LINKED ISSUE #969: Drop securedrop-workstation-bullseye base template, use a debian-12 system template
(initial attempt underway in https://github.com/freedomofpress/securedrop-workstation/tree/4_2-compat-deb12-base)

Pre-Qubes 4.2, SDW used a custom base templateVM for AppVMs, https://github.com/freedomofpress/qubes-template-securedrop-workstation . This was a maintenance burden, and now that qubes-builder is deprecated in favour of qubes-builderv2, we would need to start from scratch to stick with this approach.

An alternative would be to use the system debian-12-xfce template as a base at install time. This would reduce maintenance overhead and eliminate the 4 GiB custom template download at install time. One major downside would be a loss of control over the state of SDW AppVMs' template.

We can mitigate said downside in a couple of ways:
- Require that production SDW instances be installed on fresh Qubes installs (docs-only)
- Install and use another Qubes-provided template not installed with the OS, such as debian-12-minimal or debian-12-gnome.
- Test to see if the debian-12-xfce template is already in use at install time:
  - if so, cancel the install and instruct the user to use a fresh Qubes install
  - if not, uninstall and reinstall the template, resetting it (open question: can we do this without downloading it again?), then proceed with the install.

The last option is currently being prototyped.

### PR DESCRIPTION
**Pull request by zenmonkeykstop**
## Status

Work in progress:
- `sdw-admin --apply` command detects existing appVM dependencies on `debian-12-minimal` and alerts the user, but should also reinstall `debian-12-minimal` if necessary
- some tests failing - needs a grsec kernel with the expected identifier (including `-workstation-grsec`) and needs updates to the RPC checks.
- launcher tests are failing in CI

## Description of Changes

Fixes #969 .

## Testing

Much larger plan TK, but roughly
- [ ] Provision dev env via `make clone`, `make dev`, verify that the latter completes successfully
- [ ] reboot
- [ ] Verify that `make test` passes
- [ ] Verify that the launcher is triggered and update is successful
- [ ] Verify basic client functionality (log in, retrieve and view submissions,, reply, export/print)

## Deployment

Migration from 4.1 is not supported - fresh installs on 4.2 only.

### If you have made changes to the provisioning logic

- [ ] All tests (`make test`) pass in `dom0`

### If you have added or removed files
vi 
- [ ] I have updated `MANIFEST.in` and `rpm-build/SPECS/securedrop-workstation-dom0-config.spec`

### If documentation is required

- [ ] I have opened a PR in the [docs repo](https://github.com/freedomofpress/securedrop-workstation-docs) for these changes, or will do so later
- [ ] I would appreciate help with the documentation

### CODE REVIEW

File: Makefile
Comment (legoktm) [inline Makefile:4]:
```suggestion
PYTHON3 := $(if $(shell bash -c "command -v python3.11"), python3.11, python3)
```

File: Makefile
Comment (legoktm) [inline Makefile:41]: This seems like a blocker. Or if its just reprotest/libfaketime we can switch to the approach of doing two builds and then diffing them.

File: dom0/sd-base-template-files.sls
Comment (legoktm) [inline dom0/sd-base-template-files.sls:10]: All of these should be dependencies in -config or other packages. But I'm not sure why we're not starting with the GNOME (or even XFCE) templates to begin with, I thought that was what we had discussed initially?

File: rpm-build/SPECS/securedrop-workstation-dom0-config.spec
Comment (legoktm) [inline rpm-build/SPECS/securedrop-workstation-dom0-config.spec:56]: from the client:
```
override_dh_strip_nondeterminism:
	find ./debian/ -type f -name '*.pyc' -delete
	find ./debian/ -type d -name '__pycache__' -delete
	find ./debian/ -type f -name 'pip-selfcheck.json' -delete
	find ./debian/ -type f -name 'direct_url.json' -delete
	find ./debian/ -type f -name 'RECORD' -delete
	dh_strip_nondeterminism $@
```

let's just delete these instead of trying to fix them up

File: dom0/sd-base-template-files.sls
Comment (zenmonkeykstop) [reply dom0/sd-base-template-files.sls:10]: If we start with -gnome or -xfce, then the base template will include more than we want. The initial discussion was actually around using -minimal, I did flip to -xfce temporarily, but the resulting attack surface in terms of unused packages is too great, and removing them would be more hassle than starting with minimal and adding just what we need.

Comment (zenmonkeykstop) [reply dom0/sd-base-template-files.sls:10]: But I'm down to add them in the securedrop-workstation-config package.

**Comment by legoktm** [comment]
Summary of the dom0 test failures:

* test_sd_client_apparmor - apparently not set to enforce
* test_log_dirs_properly_named / test_logs_are_flowing - unclear what the error is, but probably related to the next two
* test_redis_service_running - it's not running
* test_sd_log_service_running - it's not running
* test_sd_proxy_rpc_spec - file doesn't exist maybe?
* test_Policies - RPC policy doesn't exist
* test_sd_workstation_template - template doesn't exist
* test_all_fedora_vms_uptodate - `Command 'sudo dnf check-update' returned non-zero exit status 100.`
* test_gpg_domain_configured - returned '' instead of sd-gpg
* test_logging_configured, and a number of related dependency-related tests all saying not installed
* test_mime_types - selecting wrong opener
* _check_kernel - known that the kernel is reporting the wrong version
* test_all_sd_vm_apt_sources - missing apt source

File: dom0/sd-dom0-qvm-rpc.sls
Comment (rocodes) [inline dom0/sd-dom0-qvm-rpc.sls:75]: (are we missing the allow rule for GpgImportKey?  `GpgImportKey * @tag:sd-client sd-gpg allow`)

File: dom0/sd-dom0-qvm-rpc.sls
Comment (rocodes) [inline dom0/sd-dom0-qvm-rpc.sls:82]: I think this is appropriate for the policy file, so that the config is set by files we control (and if `uninstall` is run it will be removed cleanly)

File: dom0/sd-logging-setup.sls
Comment (rocodes) [inline dom0/sd-logging-setup.sls:5]: nit: if we stick to ` "sd-small-{}-template".format(grains['oscodename']) %`, as in below conditional, and set it in sdvars, we will have one source of truth for debian version name

File: dom0/securedrop-check-migration
Comment (rocodes) [inline dom0/securedrop-check-migration:26]: Can we remove this, since we're starting clean on 4.2?

File: dom0/sd-dom0-files.sls
Comment (rocodes) [inline dom0/sd-dom0-files.sls:51]: I _think_ we can just use `[qvm.template_installed](https://github.com/QubesOS/qubes-mgmt-salt-dom0-qvm?tab=readme-ov-file#qvmtemplate_installed)` instead of `cmd.run`.  If memory serves, we had an issue with the qvm-template tooling because we were using the same repo to serve templates and the dom0 config rpm, and qvm-template was upset by this.

**Comment by legoktm** [comment]
I squashed it down from 41 commits to 11. 3 are annotated as FIXMES, I think we'd want to drop them before merging. I'll look into the rpm reproducibility issue tomorrow, and the shutdown VMs one can be dropped once https://github.com/freedomofpress/securedrop-workstation-ci/pull/44 is deployed. I'm not sure about the qvm-check one.

**Comment by mig5** [comment]
> the shutdown VMs one can be dropped once [freedomofpress/securedrop-workstation-ci#44](https://github.com/freedomofpress/securedrop-workstation-ci/pull/44) is deployed.

https://github.com/freedomofpress/securedrop-workstation-ci/pull/44 is now merged and deployed to the runner bastion, should now be active for all future CI runs.

File: dom0/securedrop-check-migration
Comment (zenmonkeykstop) [reply dom0/securedrop-check-migration:26]: Yup, I think so.

File: dom0/sd-sys-vms.sls
Comment (rocodes) [inline dom0/sd-sys-vms.sls:126]: Yes, if we're still going with `sd-fedora-xx-dvm` for the sys-usb dvm template, it would be friendly to clean up the old ones when we're not using them anymore.  But since we're installing clean on 4.2, users won't have an out-of-date sd-fedora dvm on their machines, so this won't be relevant til the next fedora base template upgrade.

File: files/provision-all
Comment (rocodes) [inline files/provision-all:52]: You might want `qvm-shutdown --force --wait`, but you might decide this is too niche a usecase and opt to leave it as is. The use case would be: if the user had a running appvm that used sd-whonix for networking (eg they were checking the admin interface), then sd-whonix wouldn't reboot without the `force` command.

File: files/sdw-admin.py
Comment (rocodes) [inline files/sdw-admin.py:178]: I don't think this latter part is true anymore - I think we have decided that we won't disable them, we'll just make sure all our rules take precedence before the old-style rules can even load.

File: files/sdw-admin.py
Comment (zenmonkeykstop) [reply files/sdw-admin.py:172]: This and that sticky note saying PUSH I put on a glass door in 444 Front West are my crowning achievements in life.

File: files/sdw-admin.py
Comment (zenmonkeykstop) [reply files/sdw-admin.py:178]: sdw-admin will get another pass to clean up output and add more stringent prod checks, we can bake that final decision in (and maybe check the rpc rules config as well as checking for VMs) when we do.

File: files/provision-all
Comment (legoktm) [reply files/provision-all]: I removed this commit, https://github.com/freedomofpress/securedrop-workstation-ci/pull/44 uses force so not an issue, but if we want to tweak it further, probably should go in the CI repo.

**Comment by legoktm** [comment]
I don't want to disuade people from pointing out things that can safely be removed, because I think there's a bunch, but I think we can defer them for follow-up PRs that will be trivial to merge?

----

In my testing locally, with the two commits I just pushed, the only test failures I have left are related to the kernel version. There are a few different failures remaining in CI that I wonder if they're related to the VM shutdown issue (but I can see they're being shut down in the log?)

**Comment by zenmonkeykstop** [comment]
> I don't want to disuade people from pointing out things that can safely be removed, because I think there's a bunch, but I think we can defer them for follow-up PRs that will be trivial to merge?

Yup, I agree - I'd like to get a functional 4.2 main ASAP and work on fixes/major changes/tweaks, in that order.

**Comment by legoktm** [comment]
We're down to two failing tests (excluding kernel version suffix):
```
INFO:[2024-04-04-00:02:31:790536] ======================================================================
INFO:[2024-04-04-00:02:31:790573] ERROR: test_all_fedora_vms_uptodate (test_vms_platform.SD_VM_Platform_Tests.test_all_fedora_vms_uptodate)
INFO:[2024-04-04-00:02:31:790610] Asserts that all Fedora-based templates, such as sys-net, have all
INFO:[2024-04-04-00:02:31:790647] ----------------------------------------------------------------------
INFO:[2024-04-04-00:02:31:790682] Traceback (most recent call last):
INFO:[2024-04-04-00:02:31:790718]   File "/home/user/securedrop-workstation/tests/test_vms_platform.py", line 174, in test_all_fedora_vms_uptodate
INFO:[2024-04-04-00:02:31:790752]     self._ensure_packages_up_to_date(vm, fedora=True)
INFO:[2024-04-04-00:02:31:790788]   File "/home/user/securedrop-workstation/tests/test_vms_platform.py", line 104, in _ensure_packages_up_to_date
INFO:[2024-04-04-00:02:31:790821]     stdout, stderr = vm.run(cmd)
INFO:[2024-04-04-00:02:31:790858]                      ^^^^^^^^^^^
INFO:[2024-04-04-00:02:31:790891]   File "/usr/lib/python3.11/site-packages/qubesadmin/vm/__init__.py", line 320, in run
INFO:[2024-04-04-00:02:31:790927]     raise e
INFO:[2024-04-04-00:02:31:790960]   File "/usr/lib/python3.11/site-packages/qubesadmin/vm/__init__.py", line 316, in run
INFO:[2024-04-04-00:02:31:790996]     return self.run_service_for_stdio('qubes.VMShell',
INFO:[2024-04-04-00:02:31:791029]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INFO:[2024-04-04-00:02:31:791064]   File "/usr/lib/python3.11/site-packages/qubesadmin/vm/__init__.py", line 296, in run_service_for_stdio
INFO:[2024-04-04-00:02:31:791098]     raise exc
INFO:[2024-04-04-00:02:31:791135] subprocess.CalledProcessError: Command 'sudo dnf check-update' returned non-zero exit status 100.
```

and

```
INFO:[2024-04-04-00:02:31:792645] ======================================================================
INFO:[2024-04-04-00:02:31:792681] FAIL: test_all_sd_vm_apt_sources (test_vms_platform.SD_VM_Platform_Tests.test_all_sd_vm_apt_sources)
INFO:[2024-04-04-00:02:31:792716] Test all VMs fpf apt source list iteratively.
INFO:[2024-04-04-00:02:31:792753] ----------------------------------------------------------------------
INFO:[2024-04-04-00:02:31:792787] Traceback (most recent call last):
INFO:[2024-04-04-00:02:31:792823]   File "/home/user/securedrop-workstation/tests/test_vms_platform.py", line 221, in test_all_sd_vm_apt_sources
INFO:[2024-04-04-00:02:31:792857]     self._validate_apt_sources(vm)
INFO:[2024-04-04-00:02:31:792893]   File "/home/user/securedrop-workstation/tests/test_vms_platform.py", line 83, in _validate_apt_sources
INFO:[2024-04-04-00:02:31:792927]     self.assertTrue(self.apt_url in contents)
INFO:[2024-04-04-00:02:31:792963] AssertionError: False is not true
```

I pushed some better error messages to surface which VMs are at fault. I suspect it's sys-net, etc. that have pending Fedora updates because we're not restarting them (since they don't have the sd-workstation tag). I also don't know if we can restart them, because we need a network connection to communicate with the ws-ci-bastion. If this is the case I would suggest we skip these tests if we're in the CI environment and leave a fixme for a better solution.

**Comment by zenmonkeykstop** [comment]
One option (for the package updates at least) might be to trigger the cli updater as a first/early step in the install process. For the pilot we told folk to manually update before kicking things off, but if we can do it for them so much the better.

**Comment by legoktm** [comment]
Also the VM shutdown we did in the CI repo doesn't seem sufficient, I think we might need to properly wait until they're shutdown. So we can probably ship the current FIXME commit (with a different summary) and revert it once CI is fixed.

File: dom0/sd-default-config.sls
Comment (cfm) [inline dom0/sd-default-config.sls:20]: This breaks: <https://github.com/freedomofpress/securedrop-workstation/blob/941f051f2e7fb627935dc661d01e7504b3b7c884/tests/test_vms_platform.py#L36-L37>

File: dom0/sd-dom0-files.sls
Comment (zenmonkeykstop) [reply dom0/sd-dom0-files.sls:51]: This can probably be an issue.

File: dom0/sd-logging-setup.sls
Comment (zenmonkeykstop) [reply dom0/sd-logging-setup.sls:5]: Yup that's what I meant.

**Comment by legoktm** [comment]
> I'll look into the rpm reproducibility issue tomorrow

I forgot to report back that I did investigate the reproducibility issue - the rpms are reproducible *except* when built under reprotest (I suspect this is some libfaketime interaction gone wrong). I've updated the commit/comments to reflect this and don't consider it a blocker to moving forwards.

**Review by cfm — COMMENTED** [review]
Just to stress-test things, uninstalling an existing installation on Qubes 4.2 from 4d6ca1a8cf329fccec8f2a81bca50e34ae45e0d0...5873c638d25094f734d5359c2f01005d360dde85 and then installing from here:

- [x] Provision dev env via `make clone`, `make dev`, verify that the latter completes successfully

```sh-session
[user@dom0 ~] sdw-admin --uninstall
[user@dom0 ~] cd securedrop-workstation
[user@dom0 securedrop-workstation] make clone
[user@dom0 securedrop-workstation] make dev
```

Failed because the signing key has changed per <https://github.com/freedomofpress/securedrop-workstation/pull/972>.  Apply that locally, then:

```sh-session
[user@dom0 securedrop-workstation] sdw-admin --uninstall  # our templates deleted: ✓
[user@dom0 securedrop-workstation] make clone
[user@dom0 securedrop-workstation] make dev
```

- [x] reboot

Updates, and reboot again.

- [ ] Verify that `make test` passes

Exceptions:

- review link TK
- VM kernels fail <https://github.com/freedomofpress/securedrop-workstation/blob/941f051f2e7fb627935dc661d01e7504b3b7c884/tests/test_vms_exist.py#L36>

- [x] Verify that the launcher is triggered and update is successful
- [ ] Verify basic client functionality (log in, retrieve and view submissions,, reply, export/print)

TK

File: dom0/sd-default-config.sls
Comment (zenmonkeykstop) [reply dom0/sd-default-config.sls:20]: the kernel string test failure will be addressed with next 6.6 kernels, @legoktm already fixed that in a PR pending to `kernel-builder`.

File: tests/test_qubes_rpc.py
Comment (zenmonkeykstop) [inline tests/test_qubes_rpc.py:7]: This is a good first step IMO

File: files/provision-all
Comment (zenmonkeykstop) [inline files/provision-all:51]: If we actually need a reboot, we could probably change this to a user Y/N prompt - "The system must be rebooted to complete the installation. Reboot now? Y/N"  or similar. If we don't *actually* need a reboot then kicking the VMs is fine.

**Review by zenmonkeykstop — COMMENTED** [review]
@legoktm's changes LGTM. I can't approve having created the PR tho, someone else will need to.

**Review by cfm — APPROVED** [review]
Approving based on all testing and review so far, per <https://github.com/freedomofpress/securedrop-workstation/pull/970#pullrequestreview-1990178359>.

**Pull request merged**
